### ISSUE TITLE
Backup to S3 fills temporary directory

### ISSUE THREAD

**Issue by jamesjwarren** [restic/restic#611]
## Output of `restic version`

```
restic 0.2.0 (v0.2.0-198-gd497fb6)
compiled at 2016-09-06 13:09:29 with go1.7
```
## Expected behavior

Backup files to S3
## Actual behavior

Fills temporary directory with lots of `restic-temp-pack-237545941` files and crashes due to no space left on temporary device. There are some S3 connection errors in the log. I have tried with a few directories and always get these errors on larger directories.
## Steps to reproduce the behavior

```
export AWS_ACCESS_KEY_ID=**
export AWS_SECRET_ACCESS_KEY=**
export RESTIC_PASSWORD=**
export RESTIC_REPOSITORY='s3://s3.amazonaws.com/**'

restic backup /var/qmail
restic backup /var/www
restic backup /etc
restic backup /usr
```
### Log ouput:

```
scan [/var/qmail]
scanned 5024 directories, 188254 files in 2:46
panic: client.PutObject: Put https://**.amazonaws.com/restic/data/2764cb5efda6267b69c71b26fefda81cd338523e6aaaa5d58b405652b14276af: net/http: TLS handshake timeout

goroutine 181702 [running]:
panic(0x842f20, 0xc42154a940)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
restic.(*Archiver).saveChunk(0xc420330280, 0xb2ac3f, 0x1c7cfa, 0x1e86b074000000, 0xc427eea000, 0x1c7cfa, 0x1ea000, 0xc423f14630, 0xa87e00, 0xc42222d6b0, ...)
        /tmp/restic-build-945714494/src/restic/archiver.go:165 +0x282
created by restic.(*Archiver).SaveFile
        /tmp/restic-build-945714494/src/restic/archiver.go:237 +0x31a
scan [/var/www]
scanned 40815 directories, 291865 files in 2:59
panic: client.PutObject: Put https://**.amazonaws.com/restic/data/80eaf06574c12495452eab8e98aea71952989a2480218dd03cc60080ec385080: dial tcp 54.231.133.3:443: i/o timeout

goroutine 77541 [running]:
panic(0x842f20, 0xc428667a40)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
restic.(*Archiver).saveChunk(0xc4266d2ff0, 0x35d770b, 0x1b5293, 0x1b00c3ddb00000, 0xc43198a000, 0x1b5293, 0x1ee000, 0xc420449130, 0xa87e00, 0xc424c2d3a0, ...)
        /tmp/restic-build-945714494/src/restic/archiver.go:165 +0x282
created by restic.(*Archiver).SaveFile
        /tmp/restic-build-945714494/src/restic/archiver.go:237 +0x31a
using parent snapshot 8fe72fd3
scan [/etc]
scanned 296 directories, 2222 files in 0:01
[0:28] 100.00%  1.032 MiB/s  28.904 MiB / 28.904 MiB  2518 / 2518 items  0 errors  ETA 0:00
duration: 0:28, 1.02MiB/s
snapshot 9f76e10c saved
using parent snapshot f427e217
scan [/usr]
scanned 28791 directories, 209276 files in 1:08
panic: Write: write /tmp/restic-temp-pack-398270418: no space left on device errors  ETA 7:08

goroutine 2173 [running]:
panic(0x842f20, 0xc4217c9c60)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
restic.(*Archiver).dirWorker(0xc4258fa3c0, 0xc423d053a0, 0xc420436000, 0xc4293a0f00, 0xc423d0c2a0)
        /tmp/restic-build-945714494/src/restic/archiver.go:404 +0xdeb
created by restic.(*Archiver).Snapshot
        /tmp/restic-build-945714494/src/restic/archiver.go:713 +0x751
```

**Comment by fd0** [comment]
Hi, thanks for the report. During backup, restic reads the files, splits them into chunks and bundles the chunks together into so-called "pack files". Until the upload finishes, these are stored in a temporary location like `/tmp/`. This is probably not going to change.

Do you think the amount of space that is used in `/tmp/` is unreasonable large? Do the tempfiles remain after restic has finished successfully?

From the output you pasted I can see that you ran restic multiple times for different directories, was than on purpose?

**Comment by jamesjwarren** [comment]
That makes total sense and I understand its necessary.

After this there were 27 pack files which I felt was too many, perhaps an option to limit the size of this buffer so it could wait to finish the upload of packs before creating more? Some pack files (12 * 5MB) remain after the first backup command, possibly due to the S3 connection timeout errors? Yes it was intentional to run multiple backup commands each for a different directory but doing this then exacerbates the problem as they are each leaving files behind in tmp.

Is it possible to set a custom location for the temporary storage other than `/tmp`?

**Comment by fd0** [comment]
Yes: Use the environment variable `TMPDIR`, like this:
```
$ TMPDIR=~/custom-restic-tmp-dir/ restic -r s3:... backup ...
```

**Comment by witeshadow** [comment]
Is there a mechanism to cleanup unsuccessful s3 runs?

**Comment by witeshadow** [comment]
Using a different TMPDIR via env variable did not work for me.

**Comment by fd0** [comment]
What did you do exactly? I'm pretty sure that setting `TMPDIR` as described above works just fine.

**Comment by fbkopp** [comment]
I think the program should warn you that you do not have enough space in
any way

2016-12-13 11:23 GMT-02:00 Alexander Neumann <notifications@github.com>:

**Comment by fd0** [comment]
Adding a warning for low temporary space is a good idea, thanks.

**Comment by fd0** [comment]
@witeshadow could you please check again whether setting a temporary directory via `TMPDIR` works? Which version of restic did you use exactly, what does `restic version` report?

**Comment by cakeface** [comment]
I had a similar error when backing up to S3 last night and I can't tell if it is this or a different issue.

Running `restic backup /proj/source` with `RESTIC_REPOSITORY="s3:s3.amazonaws.com/<bucket>"` I get the following error:


```
scan [/proj/source]
scanned 25654 directories, 1124955 files in 33:23
panic: client.PutObject: Put https://<bucket>.s3-eu-west-1.amazonaws.com/restic/data/a6cf009268a8a388bf700b612489317e56c514261d141de4b27e9d69232444d6: dial tcp: i/o timeout

goroutine 846786 [running]:
panic(0x87c460, 0xc4251d8500)
        /usr/lib/go/src/runtime/panic.go:500 +0x1a1
restic/archiver.(*Archiver).saveChunk(0xc4203fc280, 0x6f4b8d, 0x32ef97, 0xb02b7a400000, 0xc42333c000, 0x32ef97, 0x3c8000, 0xc420484160, 0xae29a0, 0xc420026000, ...)
        /tmp/restic-build-119647226/src/restic/archiver/archiver.go:163 +0x282
created by restic/archiver.(*Archiver).SaveFile
        /tmp/restic-build-119647226/src/restic/archiver/archiver.go:237 +0x38b
```

The actual error looks like `dial tcp: i/o timeout` which is similar to the errors above for the first 2 directories.  I don't see a ton of data in `/tmp` but there were some left over files.  My `/` partition only has about 4GB available and I am backing up around 1TB of data from an NFS drive attached.  Maybe I need a bigger temp filesystem to do the backup?

Thanks!

**Comment by cakeface** [comment]
I forgot to add, version 0.4.0 amd64 of restic on Ubuntu 14.04.

**Comment by zcalusic** [comment]
There's a fix for S3 backend that landed in repo after v0.4.0 was released. It should fix the "dial tcp: i/o timeout" error. If you can re-test with restic compiled from the current master, it might fix the network issues at least.

**Comment by fd0** [comment]
@cakeface thanks for the report, this seems to be #791, which was fixed in the master branch but is not included in the 0.4.0 release. Please try again with the master branch, if the problem persists please open a new issue.

Since there was no feedback on the original issue I'm going to close it for now.
